fix object ghost follows invisible cursor when scroll with right-click

guljam opened the pull request:
Fix issue #17221
Object ghost and tool tips following invisible cursor when scrolling with right click (like vanilla RCT2)

Fix List
- Placing attraction or track
- Footpath tool
- Land rights tool
- Guest and staff pickup
- Placing park entrance and guest start point in the scenario editor
- Staff patrol area
- Select area in cut-away view tool

Broxzier commented:
Have you tried adding this check before the `OnToolUpdate` calls are made? If placing the check before these calls works without side-effects, we could avoid all the added code inside the tool update functions.

Broxzier commented on src/openrct2-ui/windows/Footpath.cpp:
```diff
@@ -387,7 +387,11 @@ static void WindowFootpathDropdown(rct_window* w, rct_widgetindex widgetIndex, i
+    {
+        return;
+    }
+    else if (widgetIndex == WIDX_CONSTRUCT_ON_LAND)
```
`else` is pointless after an `if` that exits the function

guljam commented:
> l u

you mean like this?

if(check right mouse button pressed)
{
    return;
}
onupdate();

I've been trying to find a much higher level function, but I can't find it.

I think it is correct that the check is done inside the onupdate function.
Because the code in onupdate has to work again when the right click is released.

Broxzier commented:
What about here? https://github.com/OpenRCT2/OpenRCT2/blob/5408d053ff1cdc67bc10eb0474e878f5b2d5def5/src/openrct2-ui/input/MouseInput.cpp#L1172-L1180

guljam commented:
> What about here?
> 
> https://github.com/OpenRCT2/OpenRCT2/blob/5408d053ff1cdc67bc10eb0474e878f5b2d5def5/src/openrct2-ui/input/MouseInput.cpp#L1172-L1180

I understand the structure of game events with your code.
All files that I have modified have been reverted,
I added a right click check to the function you suggested.
As expected the object ghost doesn't follow you when you scroll, everything works fine :)

Broxzier commented on data/language/ko-KR.txt:
```diff
@@ -48,7 +48,7 @@ STR_0043    :스페이스 링
 STR_0045    :엘리베이터
 STR_0046    :수직 낙하 롤러코스터
-STR_0047    :현금지급기
+STR_0047    :현금 지급기
```
Revert all these language changes. It's probably best to rebase on develop, pick only your own commits and squash them.

Broxzier commented on distribution/changelog.txt:
```diff
@@ -21,6 +21,7 @@
 - Fix: [#17104] Changing map size does not invalidate park size.
 - Fix: [#17197] Segfault when extracting files from the GOG installer.
 - Fix: [#17205] Map generator sometimes crashes when not all standard terrain objects are available.
+- Fix: [#17245] Object ghost and tool tips follows invisible cursor when scroll viewport with right-click 
```
```suggestion
- Fix: [#17245] Object ghosts and tooltips follow invisible cursor when moving the viewport by right-click dragging.
```

Broxzier commented on src/openrct2-ui/windows/MazeConstruction.cpp:
```diff
@@ -247,6 +247,11 @@ class MazeConstructionWindow final : public Window
 
     void OnToolUpdate(rct_widgetindex widgetIndex, const ScreenCoordsXY& screenCoords) override
     {
+        if (input_get_state() == InputState::ViewportRight)
+        {
+            return;
+        }
+
```
This is no longer necessary.

Broxzier commented on src/openrct2-ui/windows/Guest.cpp:
```diff
@@ -1086,7 +1086,9 @@ void WindowGuestOverviewTextInput(rct_window* w, rct_widgetindex widgetIndex, ch
 void WindowGuestOverviewToolUpdate(rct_window* w, rct_widgetindex widgetIndex, const ScreenCoordsXY& screenCoords)
 {
     if (widgetIndex != WIDX_PICKUP)
+    {
```
Revert these unrelated changes for this PR.

Broxzier commented on src/openrct2-ui/windows/PatrolArea.cpp:
```diff
@@ -150,7 +150,9 @@ class PatrolAreaWindow final : public Window
     {
         auto mapTile = GetBestCoordsFromPos(screenCoords);
         if (!mapTile)
+        {
```
Revert these unrelated changes for this PR.

guljam replied:
I accidentally created a brach from a strange place and can't revert
I'm going to make a new PR.

Broxzier commented:
Please stop closing PRs and just fix them instead. We lose all feedback this way...

guljam commented:
@Broxzier Sorry I tried to revert the language file but it failed. I will study git more.

Broxzier commented:
Try recreating a branch with this exact name, and you might be able to reopen this PR. I'd rather we continue here that on yet another PR.

```
git branch fix-scroll-right-click-ghost f49ba7cd351a65e38896598190b972dfceb806e3
git checkout fix-scroll-right-click-ghost
```

After that, rebase this PR on develop, and only pick the commits you want to keep. Assuming `upstream` is pointing at this repo:

```
git fetch upstream
git rebase -i upstream/develop
```

In the file that opens, remove all commits but the ones you want to keep. Then save and close, then push to `origin` (assuming that `origin` is pointing at your fork).

guljam commented:
@Broxzier Before closing this PR, I did a lot of experimenting with what a revert was. As a result, many files on this branch were modified. How can I fix this?

Should I simply overwrite the newest file and then PR by modifying only a specific file?

Broxzier commented:
As mentioned in my previous comment, you can fix this by rebasing your branch and ONLY picking the commits that you want to keep. Drop all others. Then after the rebase, make the requested changes, and to make it extra nice, squash them all into one commit.

guljam commented:
@Broxzier I used the "reset --hard" command to go back to commit 0474e1c4c, edit the file and PR
In this case, is a completely new PR created again?
